src/ahp/mod.rs

```diff
@@ -359,10 +360,146 @@ impl<F: PrimeField> UnnormalizedBivariateLagrangePoly<F> for Box<dyn EvaluationD
     }
 }
 
+/// A coboundary polynomial over a domain D is a polynomial P
+/// s.t. sum_{x_in_D}(P(x)) = 0.
+/// Given a coboundary polynomial P over a domain D, a boundary
+/// polynomial is a polynomial Z s.t. P(X) = Z(g X) - Z(X) mod v_D(X)
+pub struct BoundaryPolynomial<F: PrimeField> {
+    /// The boundary polynomial.
+    poly: DensePolynomial<F>,
+    /// Evaluations of the boundary polynomial over the D domain.
+    evals: Evaluations<F>,
+}
+
+impl<F: PrimeField> Clone for BoundaryPolynomial<F> {
+    fn clone(&self) -> Self {
+        let cloned_evals = Evaluations::<F>::from_vec_and_domain(
+            self.evals.evals.clone(),
+            self.evals.domain.clone_and_box(),
+        );
+        Self {
+            poly: self.poly.clone(),
+            evals: cloned_evals
+        }
+    }
+}
+
+impl<F: PrimeField> BoundaryPolynomial<F> {
+
+    /// Construct a `self` instance from a boundary polynomial.
+    pub fn new(
+        boundary_poly: DensePolynomial<F>,
+        domain: Box<dyn EvaluationDomain<F>>
+    ) -> Result<Self, Error>
+    {
+        let poly_evals = (&boundary_poly).evaluate_over_domain_by_ref(domain);
+
+        Ok( Self { poly: boundary_poly, evals: poly_evals } )
+    }
+
+    /// Return the underlying boundary polynomial, consuming `self`.
+    pub fn polynomial(self) -> DensePolynomial<F> { self.poly }
+
+    /// Borrow the underlying boundary polynomial.
+    pub fn polynomial_ref(&self) -> &DensePolynomial<F> { &self.poly }
+
+    /// Return the evaluations over D of the boundary polynomial, consuming `self`.
+    pub fn evals(self) -> Evaluations<F> { self.evals }
+
+    /// Return the evaluations over D of the boundary polynomial, borrowing `self`.
+    pub fn evals_ref(&self) -> &Evaluations<F> { &self.evals }
+
+    /// Compute the boundary polynomial given a coboundary polynomial
+    /// evaluations `poly_evals` over the elements of a domain D.
+    pub fn from_coboundary_polynomial_evals(
+        poly_evals: Evaluations<F>
+    ) -> Result<Self, Error>
+    {
+        let domain = poly_evals.domain;
+        let evals = poly_evals.evals;
+
+        // Z(1) = 0, or any arbitrary value
+        let mut z_evals = Vec::with_capacity(evals.len());
+        z_evals.push(F::zero());
+
+        // The other coefficients of the boundary polynomial will be the cumulative sum
+        // of the evaluations of the coboundary poly over the domain, e.g.:
+        // Z(1) = 0
+        // Z(g) = Z(1) + p'(1)
+        // Z(g^2) = Z(1) + p'(1) + p'(g)
+        // ...
+        // Z(g^(|H| - 1) )= Z(1) + p(1) + p'(g) + p'(g^2) + .... + p'(g^( |H| - 2 )) ,
+        // and finally
+        // Z(g^|H|) = 0 = p'(g) + p'(g^2) + ... + p'(g^|H - 1|) + p'(g^|H|), will be excluded
+        //TODO: Prefix sum here. Parallelize ? Is it worth it ? (rayon (crossbeam too) has no parallel impl for scan())
+        let mut poly_cum_sum_evals = evals.into_iter().scan(F::zero(), |acc, x| {
```

Comment:
Possibly vector operation. How big the amount of evals?

Reply:
This cumulative sum is of a vector of size of the FFT domain in question. Typically 2^19 - 2^20 entries.
However, that process works only with field addition (not even mul), and no SM involved here. I am sceptical that moving to GPU would improve here, in particular because of the time lost in traffic.